core/src/main/java/com/datastrato/gravitino/catalog/OperationsProxy.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2024 Datastrato Pvt Ltd.
+ * This software is licensed under the Apache License version 2.
+ */
+package com.datastrato.gravitino.catalog;
+
+import com.datastrato.gravitino.catalog.rel.BaseTable;
+import com.datastrato.gravitino.rel.Table;
+import com.datastrato.gravitino.utils.PrincipalUtils;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.Collections;
+import org.apache.commons.lang3.ArrayUtils;
+
+/** Proxy wrapper on an operation class to execute operations by impersonating given user */
+public class OperationsProxy<T> implements InvocationHandler {
+
+  private final ProxyPlugin plugin;
+  private final T ops;
+
+  private OperationsProxy(ProxyPlugin plugin, T ops) {
+    this.plugin = plugin;
+    this.ops = ops;
+  }
+
+  public static <T> T createProxy(T ops, ProxyPlugin plugin) {
+    if (!(ops instanceof CatalogOperations)) {
+      throw new IllegalArgumentException("Method only supports the type of CatalogOperations");
+    }
+    return createProxyInternal(ops, plugin, ops.getClass().getInterfaces());
+  }
+
+  private static <T> T createProxyInternal(T ops, ProxyPlugin plugin, Class<?>[] interfaces) {
+    return (T)
+        Proxy.newProxyInstance(
+            ops.getClass().getClassLoader(), interfaces, new OperationsProxy(plugin, ops));
```

Comment:
Have you tested this feature in the deployed Gravitino, I'm not sure if there's a classloader issue?

Reply:
I have written an IT `ProxyCatalogHiveIT` to test this.

Reply:
It's not enough to test using mini gravitino, you should verify it manually in your local environment.

Reply:
Verified with the playground.